src/mo_frontend/typing.ml

```diff
@@ -1468,25 +1580,51 @@ and infer_exp'' env exp : T.typ =
     | _ -> ()
     end;
     let sort, ve = check_shared_pat env shared_pat in
-    let cs, tbs, te, ce = check_typ_binds env typ_binds in
+    let sort = match sort, !Mo_config.Flags.enhanced_orthogonal_persistence with
+    | T.Local T.Stable, false -> T.Local T.Flexible (* named local functions are flexible in classical mode *)
+    | _ -> sort
+    in
+    let is_async = match typ_opt with
+      | Some { it = AsyncT _; _ } -> true
+      | _ -> false
+    in
+    let is_flexible = env.named_scope = None || sort <> T.Local T.Stable || is_async in
+    let cs, tbs, te, ce = check_typ_binds env (not is_flexible) typ_binds in
     let c, ts2 = as_codomT sort typ in
     check_shared_return env typ.at sort c ts2;
     let env' = infer_async_cap (adjoin_typs env te ce) sort cs tbs (Some exp1) exp.at in
     let t1, ve1 = infer_pat_exhaustive (if T.is_shared_sort sort then local_error else warn) env' pat in
+    let parameters = new_parameters !(env'.parameters) pat in
     let ve2 = T.Env.adjoin ve ve1 in
     let ts2 = List.map (check_typ_item env') ts2 in
     typ.note <- T.seq ts2; (* HACK *)
     let codom = T.codom c (fun () -> T.Con(List.hd cs,[])) ts2 in
+    let named_scope = if is_flexible then None else enter_named_scope env name in
     if not env.pre then begin
       let env'' =
         { env' with
           labs = T.Env.empty;
           rets = Some codom;
+          named_scope;
+          captured = ref S.empty;
+          parameters = ref parameters;
           (* async = None; *) }
       in
       let initial_usage = enter_scope env'' in
       check_exp_strong (adjoin_vals env'' ve2) codom exp1;
-      leave_scope env ve2 initial_usage;
+      leave_scope env'' ve2 initial_usage;
+      assert(!closure = None);
+      closure := stable_function_closure env'' named_scope;
+      env.captured := S.union !(env''.captured) !(env'.captured);
+      (match !closure with
+      | Some Type.{ captured_variables; _ } ->
+        T.Env.iter (fun id T.{ variable_type; _ } ->
+          if not (T.stable variable_type) then
```

Comment:
I just checked the definition of T.stable and it appears to consider type parameters as stable. Why is that ok?

Reply:
This check here iterates over the variables, not type parameters. The check for stable/flexible type parameters is in `infer_call`.

`T.stable` just allows type parameters in general when EOP is on. The rule could be refined but the main check (no flexible type argument for stable parameter) is already in place at that separate location.